### PR TITLE
[Snappi] Validate PFC frames should not be counted as RX_OK, TX_OK, RX_DRP, TX_DRP

### PR DESCRIPTION
**Pull request by developfast**
### Description of PR

Summary: While receiving or sending out PFC frames, the PFC counter should be the only counter increasing on the device. The other counters are not expected to increase. The regular TX and RX counters are also polled to ensure they do not increment by the PFC count.
Fixes (issue) #6638 

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [X] Test case(new/improvement)

### Back port request
- [ ] 201911
- [X] 202012
- [X] 202205

#### What is the motivation for this PR?
PFC validation and behavior on the DUT

#### How did you do it?
1. Send PFC pause frames from TGEN to DUT, check the RX_OK, RX_DRP and PFC rx counter. Except PFC rx counter, other counters should not increase.
2. From DUT, send out PFC pause frames to neighbor, check the TX_OK, TX_DRP and PFC tx counter. Except PFC tx counter, other counter should not increase.
3. Traffic counters will increment minimally for the small amount of regular test traffic sent - this is also verified by polling the DUT. 

#### How did you verify/test it?
Tested on lab device

### CODE REVIEW

File: tests/common/snappi_tests/common_helpers.py [hunk 1]
```diff
@@ -821,6 +821,38 @@ def get_pfc_frame_count(duthost, port, priority, is_tx=False):
...
+        duthost (Ansible host instance): device under test
+        port (str): port name ex. Ethernet4
+    Returns:
+        rx_frame_count (int): Tx frame count
+    """
+    raw_out = duthost.shell("show interface counters | grep {}".format(port))['stdout']
+    rx_ok_frame_count = raw_out.split()[2]
+    rx_drp_frame_count = raw_out.split()[7]
+
+    return int(rx_ok_frame_count.replace(',', '')), int(rx_drp_frame_count.replace(',', ''))
```
Comment (neethajohn) [inline tests/common/snappi_tests/common_helpers.py:853]: use get_interfaces_counters ansible module or 'portstat -j'

File: tests/common/snappi_tests/common_helpers.py [hunk 2]
```diff
@@ -821,6 +823,42 @@ def get_pfc_frame_count(duthost, port, priority, is_tx=False):
...
+    raw_out = duthost.shell("portstat -ji {}".format(port))['stdout']
+    raw_out_stripped = re.sub(r'^.*?\n', '', raw_out, count=1)
+    raw_json = json.loads(raw_out_stripped)
+    tx_ok_frame_count = raw_json[port]["TX_OK"]
+    tx_drp_frame_count = raw_json[port]["TX_DRP"]
+
+    return int(tx_ok_frame_count.replace(',', '')), int(tx_drp_frame_count.replace(',', ''))
+
+
+def get_rx_frame_count(duthost, port):
```
Comment (baiwei0427) [inline tests/common/snappi_tests/common_helpers.py:844]: get_tx_frame_count and get_rx_frame_count are actually very similar. Can you please use a unified function to provide various TX/RX counters?

File: tests/common/snappi_tests/traffic_generation.py [hunk 3]
```diff
@@ -518,25 +520,99 @@ def verify_in_flight_buffer_pkts(duthost,
...
-            pfc_pause_rx_frames = get_pfc_frame_count(duthost, peer_port, prios[prio])
-            pytest_assert(pfc_pause_rx_frames > 0,
-                          "PFC pause frames with zero source MAC are not counted in the PFC counters")
+            pfc_pause_rx_frames = get_pfc_frame_count(duthost, peer_port, prios[prio], is_tx=False)
+            if test_traffic_pause:
+                pytest_assert(pfc_pause_rx_frames > 0,
+                              "PFC pause frames should be received and counted in RX PFC counters for priority {}"
+                              .format(prios[prio]))
+            else:
+                pytest_assert(pfc_pause_rx_frames <= PFC_FRAME_COUNT_TOL,
```
Comment (baiwei0427) [inline tests/common/snappi_tests/traffic_generation.py:559]: I do not quite understand this.

File: tests/common/snappi_tests/traffic_generation.py [hunk 4]
```diff
@@ -518,25 +520,99 @@ def verify_in_flight_buffer_pkts(duthost,
...
+
+
+def verify_tx_frame_count_dut(duthost,
+                              snappi_extra_params,
+                              tx_frame_count_deviation=0.01,
+                              tx_drop_frame_count_tol=5):
+    """
+    Verify correct frame count for tx frames on the DUT
+    (OK and DROPS) when the traffic is expected to be paused on the DUT.
+    DUT is polled after it stops receiving PFC pause frames from TGEN.
```
Comment (baiwei0427) [inline tests/common/snappi_tests/traffic_generation.py:570]: When you poll DUT, has DUT drained all the packets in the buffer?

File: tests/common/snappi_tests/traffic_generation.py [hunk 5]
```diff
@@ -518,25 +520,99 @@ def verify_in_flight_buffer_pkts(duthost,
...
+        duthost (obj): DUT host object
+        snappi_extra_params (SnappiTestParams obj): additional parameters for Snappi traffic
+        rx_frame_count_deviation (float): deviation for rx frame count (default to 1%)
+        rx_drop_frame_count_tol (int): tolerance for tx drop frame count
+    Returns:
+
+    """
+    dut_port_config = snappi_extra_params.base_flow_config["dut_port_config"]
+    pytest_assert(dut_port_config is not None, 'Flow port config is not provided')
+    tgen_tx_frames = snappi_extra_params.test_tx_frames
```
Comment (baiwei0427) [inline tests/common/snappi_tests/traffic_generation.py:608]: no tgen

**Review by baiwei0427 — CHANGES_REQUESTED** [review]

File: tests/common/snappi_tests/traffic_generation.py [hunk 6]
```diff
@@ -7,14 +7,16 @@
...
     get_lossless_buffer_size, get_pg_dropped_packets,\
-    sec_to_nanosec, get_pfc_frame_count, packet_capture
+    sec_to_nanosec, get_pfc_frame_count, packet_capture, get_tx_frame_count, get_rx_frame_count
 from tests.common.snappi_tests.port import select_ports, select_tx_port
 from tests.common.snappi_tests.snappi_helpers import wait_for_arp
 
 logger = logging.getLogger(__name__)
 
 SNAPPI_POLL_DELAY_SEC = 2
 CONTINUOUS_MODE = -5
```
[the remark this answers is not in the document]
Comment (developfast) [reply tests/common/snappi_tests/traffic_generation.py:17]: It's an arbitrary number that isn't used anywhere else - continuous mode means the traffic flows constantly without any fixed duration or fixed packets at some percentage rate

File: tests/common/snappi_tests/traffic_generation.py (hunk 4 above)
Comment (developfast) [reply tests/common/snappi_tests/traffic_generation.py:570]: Yes

File: tests/common/snappi_tests/traffic_generation.py [hunk 7]
```diff
@@ -518,25 +519,101 @@ def verify_in_flight_buffer_pkts(duthost,
...
-            pytest_assert(pfc_pause_rx_frames > 0,
-                          "PFC pause frames with zero source MAC are not counted in the PFC counters")
+            pfc_pause_tx_frames = get_pfc_frame_count(duthost, peer_port, prios[prio], is_tx=True)
+            if test_traffic_pause:
+                pytest_assert(pfc_pause_tx_frames > 0,
+                              "PFC pause frames should be transmitted and counted in TX PFC counters for priority {}"
+                              .format(prios[prio]))
+            else:
+                # PFC pause frames should not be transmitted when test traffic is not paused
+                pytest_assert(pfc_pause_tx_frames <= PFC_FRAME_COUNT_TOL,
```
Comment (baiwei0427) [inline tests/common/snappi_tests/traffic_generation.py:548]: have you ever seen pfc_pause_tx_frames >0 in this case?

Comment (developfast) [reply tests/common/snappi_tests/traffic_generation.py:548]: No - this is just there as a back stopper. If you want me to lower the threshold to 2 or 3, that is also possible. It is currently set to 5.

Comment (developfast) [reply tests/common/snappi_tests/traffic_generation.py:548]: Reset to 0

Comment (baiwei0427) [reply tests/common/snappi_tests/traffic_generation.py:548]: > Reset to 0

Can you please remove PFC_FRAME_COUNT_TOL?

File: tests/common/snappi_tests/traffic_generation.py [hunk 8]
```diff
@@ -518,25 +519,101 @@ def verify_in_flight_buffer_pkts(duthost,
...
         snappi_extra_params (SnappiTestParams obj): additional parameters for Snappi traffic
     Returns:
 
     """
     dut_port_config = snappi_extra_params.base_flow_config["dut_port_config"]
     pytest_assert(dut_port_config is not None, 'Flow port config is not provided')
 
-    for peer_port, prios in dut_port_config[1].items():
+    for peer_port, prios in dut_port_config[0].items():  # TX PFC pause frames
         for prio in range(len(prios)):
```
Comment (baiwei0427) [inline tests/common/snappi_tests/traffic_generation.py:540]: for prio in prios?

File: tests/common/snappi_tests/traffic_generation.py [hunk 9]
```diff
@@ -518,25 +519,101 @@ def verify_in_flight_buffer_pkts(duthost,
...
+                pytest_assert(pfc_pause_tx_frames > 0,
+                              "PFC pause frames should be transmitted and counted in TX PFC counters for priority {}"
+                              .format(prios[prio]))
+            else:
+                # PFC pause frames should not be transmitted when test traffic is not paused
+                pytest_assert(pfc_pause_tx_frames <= PFC_FRAME_COUNT_TOL,
+                              "PFC pause frames should not be transmitted and counted in TX PFC counters")
+
+    for peer_port, prios in dut_port_config[1].items():  # RX PFC pause frames
+        for prio in range(len(prios)):
```
Comment (baiwei0427) [inline tests/common/snappi_tests/traffic_generation.py:552]: for prio in prios

**Review by neethajohn — APPROVED** [review]

**Pull request merged**